engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/BigNumberTest.java

```diff
@@ -177,4 +179,39 @@ public void doubleBigInteger() throws Exception {
     var fourtyTwo = assertMul(6.0, new BigInteger("7"));
     assertEquals(42, fourtyTwo.asInt());
   }
+
+  @Test
+  public void everyValueSmallerThanIntegerMaxVal_IsPrimitiveInt() {
+    var almostMaxInt = Integer.toString(Integer.MAX_VALUE - 1);
+    var intVal = ContextUtils.evalModule(ctx, "main = " + almostMaxInt);
+    assertThat("Is a number", intVal.isNumber(), is(true));
+    assertThat("Fits in int", intVal.fitsInInt(), is(true));
+    assertThat("Fits in long", intVal.fitsInLong(), is(true));
+    assertThat("Fits in double", intVal.fitsInDouble(), is(true));
+    assertThat("Fits in big int", intVal.fitsInBigInteger(), is(true));
+  }
+
+  @Test
+  public void everyValueSmallerThanLongMaxVal_IsPrimitiveLong() {
+    var almostMaxLong = Long.toString(Long.MAX_VALUE - 1);
+    var longVal = ContextUtils.evalModule(ctx, "main = " + almostMaxLong);
+    assertThat("Is a number", longVal.isNumber(), is(true));
+    assertThat("Does not fit in int", longVal.fitsInInt(), is(false));
+    assertThat("Fits in long", longVal.fitsInLong(), is(true));
+    assertThat("Does not fit in double (but could)", longVal.fitsInDouble(), is(false));
```

Comment:
This is surprising. A value that fits in `Long`, but does not fit in `Double` (according to Enso) ??

Note that all of these 3 tests pass in develop

Reply:
> A value that fits in Long, but does not fit in Double (according to Enso) ??

Every long value bigger than 2^52 that is not power of 2 doesn't fit into `double`. It doesn't fit because _it would loose precision_.

Reply:
@GregoryTravis is our export on `BigInteger`, `BigDecimal`, `Float`, `Integer` conversions.

Reply:
Fixed the issue by [not converting double to BigInt every time](1d012215e4ebdb42e13fc8eb98f96d1ca2f6ca27)